content/posts/2024-08-29-gsoc-three-way-comparison.md

```diff
@@ -0,0 +1,90 @@
+---
+author: "Volodymyr Vasylkun (Poseydon)"
+date: "2024-08-29"
+tags: ["GSoC", "optimization", "intrinsics"]
+title: "GSoC 2024: 3-way comparison intrinsics"
+---
+
+Hello everyone! My name is Volodymyr, and in this post I would like to talk about the project I have been working on for the past couple of months as part of Google Summer of Code 2024. The aim of the project was to introduce 3-way comparison intrinsics to LLVM IR and add a decent level of optimizations for them.
+
+# Background
+
+Three-way comparison is an operation present in many high-level languages, such as C++ and its starhip operator or Rust and the `Ord` trait. It operates on two values for which there is a defined comparison operation and returns `-1` if the first operand is less than the second, `0` if they are equal, and `1` otherwise. At the moment, compilers that use LLVM express this operation using different sequences of instructions which are optimized and lowered individually rather than as a single operation. Adding an intrinsic for this operation would therefore help us generate better machine code on some targets, as well as potentially optimize patterns in the middle-end that we didn't optimize before.
```

Comment:
```suggestion
Three-way comparison is an operation present in many high-level languages, such as C++ and its starship operator or Rust and the `Ord` trait. It operates on two values for which there is a defined comparison operation and returns `-1` if the first operand is less than the second, `0` if they are equal, and `1` otherwise. At the moment, compilers that use LLVM express this operation using different sequences of instructions which are optimized and lowered individually rather than as a single operation. Adding an intrinsic for this operation would therefore help us generate better machine code on some targets, as well as potentially optimize patterns in the middle-end that we didn't optimize before.
```

Reply:
It is called `spaceship operator`.